core/templates/dev/head/domain/subtopic_viewer/SubtopicDataObjectFactory.ts

```diff
@@ -0,0 +1,70 @@
+// Copyright 2019 The Oppia Authors. All Rights Reserved.
+//
```

Comment:
I believe this is the expected way, this is the pattern that I have seen at most of the places.

Reply:
@Showtim3, I don't see any naming convention for the "Factory" files over [here](https://github.com/oppia/oppia/wiki/The-File-Naming-Convention-and-Directory-Structure#the-naming-convention), can you please add one? 

@bansalnitish, Can you also confirm this?

Reply:
@Showtim3, I don't see any naming convention for the "Factory" files over [here](https://github.com/oppia/oppia/wiki/The-File-Naming-Convention-and-Directory-Structure#the-naming-convention), can you please add one? 

@bansalnitish, Can you also confirm this?

Reply:
I'll wait for the approval of @bansalnitish , then I'll add the convention for factory files.

Reply:
Yup looks fine to me -- as said same pattern is followed.

Reply:
@bansalnitish, is there any specific reason we follow this (inconsistent)pattern for the factory services?

Reply:
No -- I'm not sure

Reply:
@bansalnitish, In future after migration are we going to have the `SubtopicDataObjectFactory` service or just the `SubtopicData` class with createFromBackendDict as a static class? (Asking this as this may affect the file name.)

@seanlip, Do you have any comments on the filename pattern? (I think we should have a consistent filename in the frontend.)

Reply:
As long as we use a single convention in the overall codebase, I'm probably fine with anything.

I think SubtopicDataObjectFactory.ts seems like a reasonable filename; what alternative were you considering?

Reply:
We follow the following patterns for filename:

For frontend services: <service-name>.service.ts
For frontend directive: <directive-nmae>.directive.ts
(As documented [here](https://github.com/oppia/oppia/wiki/The-File-Naming-Convention-and-Directory-Structure#the-naming-convention).)

So considering this I was expecting the filename for factory services to be `<class-name>.factory.ts`.

Reply:
Oh, I see. Yes, I think that subtopic-data.factory.ts makes sense, then.

Reply:
> I believe this is the expected way, this is the pattern that I have seen at most of the places.

> Yup looks fine to me -- as said same pattern is followed.

Btw, I'm a bit concerned about this, if the existing convention is inconsistent -- it seems like this is already causing developer confusion. Would it be possible to file a bug listing the files that should be renamed?

Reply:
> @bansalnitish, In future after migration are we going to have the SubtopicDataObjectFactory service or just the SubtopicData class with createFromBackendDict as a static class? (Asking this as this may affect the file name.)

I'm not sure whether in future we are going to have an injectable service for the domain object factory or it will just be a class with `createFromBackendDict` as a class/static method. I think this will also help us finalize the name and properly structure our domain objects for Angular. 

@bansalnitish & @Showtim3, can you please reply to this?  I can create an issue once you can reply/resolve my above comment.

Reply:
I've raised the above question as [this](https://github.com/oppia/oppia/pull/8121/files#diff-e45c7f2cecc5e8292d3ef6b83f1f0bb7R39-R43) (I see similar patterns in other places as well.) doesn't look good to me as for creating a domain object I've to create new service and inject so many factoryServices, though the class can handel it independently. (I'm not much familiar with Angular structure, so sorry if I'm raising odd questions.)  

Also, why we are doing [this](https://github.com/oppia/oppia/pull/8121/files#diff-f6da2ad940ed2db9b011f3aef27ac8e6R65-R67) if we are not going to inject it in as an AngularJS dependency?

Reply:
Ok so @DubeySandeep We should rename all the factory files as to ``<class-name>.factory.ts`` ? Yeah it'd be better. 
We have a tracker sheet that contains all the services. That might be helpful.

Reply:
> I'm not sure whether in future we are going to have an injectable service for the domain object factory or it will just be a class with createFromBackendDict as a class/static method. I think this will also help us finalize the name and properly structure our domain objects for Angular.

For each object factory we will have two classes like `SubtopicData` and `SubtopicDataObjectFactory`, wherein the `SubtopicDataObjectFactory` will have the injectable decorator and this class will have createFromBackend function.

Let me know if you need any other information related to this. +1 for `<class-name>.factory.ts`!